Refactor setup.sh to install bower and node dependencies from composer itself

monishdeb opened the pull request:
Overview
----------------------------------------
This PR shifts the code which installs bower and node dependencies, into a separate shell script file ```tools/scripts/composer/download-dependencies.sh```. Later executed via composer on post install.

Before
----------------------------------------
Bower and node dependencies are installed via distmaker or on manual CLI execution.

After
----------------------------------------
Bower and node dependencies are installed via composer.

monishdeb commented:
ping @totten @dsnopek @seamuslee001 @colemanw

dsnopek commented:
Ooo, this is a nice incremental improvement! The user still needs to have bower installed, but it removes, some process that they need to know, which is good.

What happens if the user doesn't have bower? Is the error message helpful? If it fails because they don't have bower, and then install it, will re-running `composer install` just work?

I can do some testing later, but probably not today.

monishdeb commented:
@dsnopek yes tried that scenario. In my setup to in order to replicate it which is what if ```bower``` or ```nodejs``` command is not found, I simply put wrong name say ```BOWER=$(pickcmd node_modules/bower/bin/bowerr bowerr);``` it simply proceeds with an error message
```
++ pickcmd node_modules/bower/bin/bowerr bowerr
++ for name in '"$@"'
++ which node_modules/bower/bin/bowerr
++ for name in '"$@"'
++ which bowerr
++ echo 'ERROR: Failed to find any of these commands: node_modules/bower/bin/bowerr' bowerr
++ exit 1
+ BOWER='ERROR: Failed to find any of these commands: node_modules/bower/bin/bowerr bowerr'
+ '[' -f 'ERROR: Failed to find any of these commands: node_modules/bower/bin/bowerr bowerr' ']'
+ '[' -n '' ']'
+ '[' '!' -f bower_components/.setupsh.ts ']'
+ '[' bower.json -nt bower_components/.setupsh.ts ']'
+ '[' -f bower_components/.setupsh.ts ']'
+ rm -f bower_components/.setupsh.ts
+ ERROR: Failed to find any of these commands: node_modules/bower/bin/bowerr bowerr install
tools/scripts/composer/download-dependencies.sh: line 33: ERROR:: command not found
+ touch bower_components/.setupsh.ts
```

In here ```composer install``` won't be hampered because the particular shell script is called on [post install](https://github.com/civicrm/civicrm-core/pull/12115/files#diff-b5d0ee8c97c7abd7e3fa29b9a27d1780R67) plus the new shell script introduced here is itself won't hinder the install process.

seamuslee001 commented:
Question for composer experts @dsnopek @monishdeb @totten will the scripts in the post-install-cmd etc be run even when CiviCRM is a dependency just looking at https://groups.google.com/forum/#!topic/composer-dev/6eL32Rmq3-w suggests maybe not

dsnopek commented:
No, I don't think they'll be run when a project depends on civicrm-core. But breaking it out into a script could allow putting that script into a "project template" that people can use for their Drupal 8 & CiviCRM sites via `composer create-project` which is what I'm hoping will eventually replace my instructions :-)

totten commented:
Background:

1. In [CRM-17652](https://issues.civicrm.org/jira/browse/CRM-17652?focusedCommentId=105842&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-105842
) @dsnopek raised a few approaches to the bower/composer situation.
2. In [dev/drupal#8](https://lab.civicrm.org/dev/drupal/issues/8), the issue is just a stub that's flagged as `needs:planning` -- i.e. it doesn't have a specification, and my expectation was that we needed to have more discussion/evaluation before settling on a design.

Now, how to interpret this PR... It depends a lot on how it's labelled/presented:

* (a) If presented as a refactor that just tries to break-down a big file/function (`setup.sh`) by extracting a smaller file/function (`download-dependencies.sh`), then that goal seems nice enough, and the changes would meet the expectation.
* (b) If presented as a solution for the composer/bower situation, then it feels unsatisfying. It doesn't talk enough about the downstream admin experience, about the trade-offs of the approach or why those are better than the others, about what happens when extension authors try to imitate core's approach, etc.

If you want to reframe the issue as (a), then I think it'll be easier to review/merge. If you want to keep it under frame (b), then we're in for a longer ride.

monishdeb commented:
Sorry, I took a wrong approach despite the issue is labeled with ```needs:planning``` and I should have discussed it first for finalizing the spec. I have changed the PR title and description, removed the issue reference, which is not appropriate of what the patch actually tries to achieve.

monishdeb commented:
This is a not a proper fix to address the main issue here, not even this refactoring change will be helpful in any way to load bower dependencies via composer. Hence I am closing this PR of mine.

dsnopek commented:
As an incremental change, I don't think it's a terrible idea? We'll still need something to be run from actual Drupal 8 composer projects, and having this functionality broken out into its own script would make that easier, I think.

monishdeb commented:
@dsnopek yes it can solve the problem for now but after I did some testing with [Asset Packagist](https://asset-packagist.org/) way I found a possibility to load all dependencies on desired path ```bower_compoenents``` without the need of any post-update shellscript. But later realized that it will be a long leap to cover as there are some limitation at hand as described in https://lab.civicrm.org/dev/drupal/issues/8#note_4752
